oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/IntrospectionClaimNames.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.resource.authentication;
+
+/**
+ * The names of the &quot;Introspection Claims&quot; defined by an
+ * <a target="_blank" href="https://tools.ietf.org/html/rfc7662#section-2.2">Introspection Response</a>.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public interface IntrospectionClaimNames {
```

Comment:
> Instead perhaps it makes more sense to have concrete methods

I agree and was more envisioning adopting the same pattern as `JwtClaimAccessor`, `IdTokenClaimAccessor`, and `StandardClaimAccessor`. Defining the constant only is the first step. We still need something like `IntrospectionClaimAccessor`

Reply:
What is the use case for end users actually using this class? As a user, I don't care about accessing "introspected" attributes, per se. JWT is a structure, so it makes sense for me to have a `JwtClaimAccessor`, e.g. "I want to access claims this through the JWT-defined structure". "Introspection" is not a structure.

I think it's sensible to have a common place for attributes for classes that need to commonly refer to introspection-defined properties, but it isn't really the end user that needs them.

I don't yet see the need for an Introspection-specific claim accessor, but I think we could add this later if the need becomes apparent.

Reply:
The user shouldn't have to care that the token was introspected. They care that the claims are available and trusted, both of which are implied by their controller method being invoked (and Spring Security being correctly configured). That they were obtained via an introspection endpoint is immaterial to the user.

For example, if I want to get the `aud` attribute in my controller, it's more flexible to do one of these:

```java
token.getAttributes("aud")
// or
token.getAttributes(CommonAttributeNames.AUDIENCE)
// or
token.getAttributes(JwtClaimNames.AUDIENCE) // see my reasoning below
```

Or, so long as the object hierarchy isn't bound to the way it was validated:

```java 
token.getAudience()
```

An approach that makes sense to me is to write my code this way:

```java
@GetMapping("/protected")
public String get(StructuredOAuth2Token token) {
    JwtClaimsAccessor accessor = () -> token.getAttributes();
    List<String> audience = accessor.getAudience();
    // ...
}
```

This is sensible because I'm accessing the JWT-like attributes, which is the same thing that the spec does (emphasis mine):

> aud
      OPTIONAL.  Service-specific string identifier or list of string
      identifiers representing the intended audience for this token, **as
      defined in JWT**

And to be clear, I'm not saying that the user shouldn't ever care. Like you said, Introspection has attributes that are specific to its defined structure, and should I care to retrieve attributes that are Introspection-specific, it's reasonable that the API would indiciate that I've crossed into Introspection-specific territory.

For example, if the user wanted to do `token.getClientId()`, the only way we'd be able to do that is through an Introspection-specific hierarchy. I don't see this as a common attribute to retrieve, but I agree that it's valuable for the sake of completeness, so I've created a ticket to add `IntrospectionClaimAccessor` that we can prioritize separately.

Reply:
This conversation is resolved via https://github.com/spring-projects/spring-security/issues/6498